# 02 — Comprensión de los datos (Data Understanding)

**Fase CRISP-DM:** Comprensión de los datos.

**Lee de:** `data/raw/` (arabica_data_cleaned.csv, robusta_data_cleaned.csv — [Coffee Quality Institute database](https://github.com/jldbc/coffee-quality-database), sin modificar).

**Produce:** `reports/figures/` (pendiente, ver TODO al final) y deja documentados los hallazgos que alimentan la fase 03 (`data/interim/` aún no recibe nada nuevo en esta fase: no se aplicó ninguna transformación, solo exploración).

Este notebook reorganiza y ejecuta, usando funciones de `src/`, la exploración que ya existía en el notebook original (`notebooks/_original/Proyecto_Final_Machine_Learning.ipynb`, celdas 1 a 9). No se agregó ningún análisis nuevo.

In [1]:
import sys
sys.path.insert(0, "..")  # para poder importar src/ desde notebooks/

from src.config import load_config
from src.data.load_data import load_raw_data
from src.utils import null_summary

config = load_config()
config["project"]["random_seed"]

42

## 2.1 Dataset arabica

In [2]:
df_arabica = load_raw_data("arabica", config)
df_arabica.head(10)

,Unnamed: 0,Species,Owner,Country.of.Origin,Farm.Name,Lot.Number,Mill,ICO.Number,Company,Altitude,...,Color,Category.Two.Defects,Expiration,Certification.Body,Certification.Address,Certification.Contact,unit_of_measurement,altitude_low_meters,altitude_high_meters,altitude_mean_meters
0,1,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,0,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0
1,2,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,1,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0
2,3,Arabica,grounds for health admin,Guatemala,"san marcos barrancas ""san cristobal cuch",NaN,NaN,NaN,NaN,1600 - 1800 m,...,NaN,0,"May 31st, 2011",Specialty Coffee Association,36d0d00a3724338ba7937c52a378d085f2172daa,0878a7d4b9d35ddbf0fe2ce69a2062cceb45a660,m,1600.0,1800.0,1700.0
3,4,Arabica,yidnekachew dabessa,Ethiopia,yidnekachew dabessa coffee plantation,NaN,wolensu,NaN,yidnekachew debessa coffee plantation,1800-2200,...,Green,2,"March 25th, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1800.0,2200.0,2000.0
4,5,Arabica,metad plc,Ethiopia,metad plc,NaN,metad plc,2014/2015,metad agricultural developmet plc,1950-2200,...,Green,2,"April 3rd, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1950.0,2200.0,2075.0
5,6,Arabica,ji-ae ahn,Brazil,NaN,NaN,NaN,NaN,NaN,NaN,...,Bluish-Green,1,"September 3rd, 2014",Specialty Coffee Institute of Asia,726e4891cf2c9a4848768bd34b668124d12c4224,b70da261fcc84831e3e9620c30a8701540abc200,m,NaN,NaN,NaN
6,7,Arabica,hugo valdivia,Peru,NaN,NaN,hvc,NaN,richmond investment-coffee department,NaN,...,Bluish-Green,0,"September 17th, 2013",Specialty Coffee Institute of Asia,726e4891cf2c9a4848768bd34b668124d12c4224,b70da261fcc84831e3e9620c30a8701540abc200,m,NaN,NaN,NaN
7,8,Arabica,ethiopia commodity exchange,Ethiopia,aolme,NaN,c.p.w.e,010/0338,NaN,1570-1700,...,NaN,0,"September 2nd, 2011",Ethiopia Commodity Exchange,a176532400aebdc345cf3d870f84ed3ecab6249e,61bbaf6a9f341e5782b8e7bd3ebf76aac89fe24b,m,1570.0,1700.0,1635.0
8,9,Arabica,ethiopia commodity exchange,Ethiopia,aolme,NaN,c.p.w.e,010/0338,NaN,1570-1700,...,NaN,0,"September 2nd, 2011",Ethiopia Commodity Exchange,a176532400aebdc345cf3d870f84ed3ecab6249e,61bbaf6a9f341e5782b8e7bd3ebf76aac89fe24b,m,1570.0,1700.0,1635.0
9,10,Arabica,diamond enterprise plc,Ethiopia,tulla coffee farm,NaN,tulla coffee farm,2014/15,diamond enterprise plc,1795-1850,...,Green,4,"March 29th, 2016",METAD Agricultural Development plc,309fcf77415a3661ae83e027f7e5f05dad786e44,19fef5a731de2db57d16da10287413f5f99bc2dd,m,1795.0,1850.0,1822.5


In [3]:
df_arabica.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1311 entries, 0 to 1310
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             1311 non-null   int64  
 1   Species                1311 non-null   object 
 2   Owner                  1304 non-null   object 
 3   Country.of.Origin      1310 non-null   object 
 4   Farm.Name              955 non-null    object 
 5   Lot.Number             270 non-null    object 
 6   Mill                   1001 non-null   object 
 7   ICO.Number             1163 non-null   object 
 8   Company                1102 non-null   object 
 9   Altitude               1088 non-null   object 
 10  Region                 1254 non-null   object 
 11  Producer               1081 non-null   object 
 12  Number.of.Bags         1311 non-null   int64  
 13  Bag.Weight             1311 non-null   object 
 14  In.Country.Partner     1311 non-null   object 
 15  Harv

**Hallazgo:** `df_arabica` tiene 1311 filas y 44 columnas. No se observan tipos de dato inconsistentes a simple vista (mezcla esperada de `object`, `int64` y `float64`).

In [4]:
print("Filas duplicadas:", df_arabica.duplicated().sum())

Filas duplicadas: 0


In [5]:
null_summary(df_arabica)

,cantidad_nulos,porcentaje
Lot.Number,1041,79.41
Farm.Name,356,27.15
Mill,310,23.65
Color,267,20.37
Producer,230,17.54
altitude_low_meters,227,17.32
altitude_high_meters,227,17.32
altitude_mean_meters,227,17.32
Altitude,223,17.01
Company,209,15.94


**Hallazgo:** no hay filas completamente duplicadas. Las columnas con más nulos son `Lot.Number` (79.41%), `Farm.Name` (27.15%), `Mill` (23.65%), `Color` (20.37%), `Producer` (17.54%) y las tres columnas de altitud (~17.32%). Esto es información real extraída del dataset; la estrategia de imputación/descarte de estas columnas **no se ha decidido todavía** y queda para la fase 03 (ver TODO).

## 2.2 Dataset robusta

In [6]:
df_robusta = load_raw_data("robusta", config)
df_robusta.head(10)

,Unnamed: 0,Species,Owner,Country.of.Origin,Farm.Name,Lot.Number,Mill,ICO.Number,Company,Altitude,...,Color,Category.Two.Defects,Expiration,Certification.Body,Certification.Address,Certification.Contact,unit_of_measurement,altitude_low_meters,altitude_high_meters,altitude_mean_meters
0,1,Robusta,ankole coffee producers coop,Uganda,kyangundu cooperative society,NaN,ankole coffee producers,0,ankole coffee producers coop,1488,...,Green,2,"June 26th, 2015",Uganda Coffee Development Authority,e36d0270932c3b657e96b7b0278dfd85dc0fe743,03077a1c6bac60e6f514691634a7f6eb5c85aae8,m,1488.0,1488.0,1488.0
1,2,Robusta,nishant gurjer,India,sethuraman estate kaapi royale,25,sethuraman estate,14/1148/2017/21,kaapi royale,3170,...,NaN,2,"October 31st, 2018",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,3170.0,3170.0,3170.0
2,3,Robusta,andrew hetzel,India,sethuraman estate,NaN,NaN,0000,sethuraman estate,1000m,...,Green,0,"April 29th, 2016",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,1000.0,1000.0,1000.0
3,4,Robusta,ugacof,Uganda,ugacof project area,NaN,ugacof,0,ugacof ltd,1212,...,Green,7,"July 14th, 2015",Uganda Coffee Development Authority,e36d0270932c3b657e96b7b0278dfd85dc0fe743,03077a1c6bac60e6f514691634a7f6eb5c85aae8,m,1212.0,1212.0,1212.0
4,5,Robusta,katuka development trust ltd,Uganda,katikamu capca farmers association,NaN,katuka development trust,0,katuka development trust ltd,1200-1300,...,Green,3,"June 26th, 2015",Uganda Coffee Development Authority,e36d0270932c3b657e96b7b0278dfd85dc0fe743,03077a1c6bac60e6f514691634a7f6eb5c85aae8,m,1200.0,1300.0,1250.0
5,6,Robusta,andrew hetzel,India,NaN,NaN,(self),NaN,"cafemakers, llc",3000',...,Green,0,"February 28th, 2013",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,3000.0,3000.0,3000.0
6,7,Robusta,andrew hetzel,India,sethuraman estates,NaN,NaN,NaN,cafemakers,750m,...,Green,0,"May 15th, 2015",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,750.0,750.0,750.0
7,8,Robusta,nishant gurjer,India,sethuraman estate kaapi royale,7,sethuraman estate,14/1148/2017/18,kaapi royale,3140,...,Bluish-Green,0,"October 25th, 2018",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,3140.0,3140.0,3140.0
8,9,Robusta,nishant gurjer,India,sethuraman estate,RKR,sethuraman estate,14/1148/2016/17,kaapi royale,1000,...,Green,0,"August 17th, 2017",Specialty Coffee Association,ff7c18ad303d4b603ac3f8cff7e611ffc735e720,352d0cf7f3e9be14dad7df644ad65efc27605ae2,m,1000.0,1000.0,1000.0
9,10,Robusta,ugacof,Uganda,ishaka,NaN,nsubuga umar,0,ugacof ltd,900-1300,...,Green,6,"August 5th, 2015",Uganda Coffee Development Authority,e36d0270932c3b657e96b7b0278dfd85dc0fe743,03077a1c6bac60e6f514691634a7f6eb5c85aae8,m,900.0,1300.0,1100.0


In [7]:
df_robusta.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 44 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             28 non-null     int64  
 1   Species                28 non-null     object 
 2   Owner                  28 non-null     object 
 3   Country.of.Origin      28 non-null     object 
 4   Farm.Name              25 non-null     object 
 5   Lot.Number             6 non-null      object 
 6   Mill                   20 non-null     object 
 7   ICO.Number             17 non-null     object 
 8   Company                28 non-null     object 
 9   Altitude               25 non-null     object 
 10  Region                 26 non-null     object 
 11  Producer               26 non-null     object 
 12  Number.of.Bags         28 non-null     int64  
 13  Bag.Weight             28 non-null     object 
 14  In.Country.Partner     28 non-null     object 
 15  Harvest.

**Hallazgo:** `df_robusta` tiene solo 28 filas frente a las 1311 de `df_arabica` — un desbalance severo (~47:1) entre especies. Esto es coherente con que el notebook original haya importado herramientas de `imblearn` (`SMOTE`, `RandomUnderSampler`, `SMOTEENN`), aunque el notebook original nunca llegó a usarlas ni definió si el desbalance a tratar es este (arabica vs. robusta) u otro (por ejemplo, la distribución de la futura variable "café de especialidad"). **Esta decisión queda pendiente y debe documentarse en `01_comprension_del_negocio.ipynb` / `03_preparacion_de_los_datos.ipynb`.**

In [8]:
print("Filas duplicadas:", df_robusta.duplicated().sum())

Filas duplicadas: 0


In [9]:
null_summary(df_robusta)

,cantidad_nulos,porcentaje
Variety,25,89.29
Lot.Number,22,78.57
Processing.Method,18,64.29
ICO.Number,11,39.29
Mill,8,28.57
altitude_mean_meters,3,10.71
Altitude,3,10.71
Farm.Name,3,10.71
altitude_low_meters,3,10.71
altitude_high_meters,3,10.71


**Hallazgo:** no hay duplicados. Las columnas con más nulos son `Variety` (89.29%), `Lot.Number` (78.57%), `Processing.Method` (64.29%), `ICO.Number` (39.29%) y `Mill` (28.57%) — proporciones aún más altas que en arabica, consistente con el tamaño muestral tan pequeño de este subconjunto.

## 2.3 Variable candidata a objetivo: `Total.Cup.Points`

In [10]:
df_arabica["Total.Cup.Points"].describe()

count    1311.000000
mean       82.115927
std         3.515761
min         0.000000
25%        81.170000
50%        82.500000
75%        83.670000
max        90.580000
Name: Total.Cup.Points, dtype: float64

**Hallazgo:** la media de `Total.Cup.Points` en arabica es 82.12 (std 3.52), con un máximo de 90.58, pero un **mínimo de 0.0** — muy por debajo del percentil 25 (81.17). Un puntaje de 0 no es un puntaje de catación real (la escala SCA va de 0 a 100 mediante la suma de subpuntajes de aroma, sabor, acidez, etc., y en la práctica los lotes evaluados rara vez bajan de 60); esto sugiere fuertemente un **registro corrupto o un lote descalificado/no evaluado**, no una observación válida de baja calidad. **Esta fila debe identificarse y tratarse explícitamente (no simplemente eliminarse sin justificar) en la fase 03**, ya que de dejarse tal cual distorsionaría cualquier modelo o umbral que dependa de `Total.Cup.Points`.

Contexto de negocio (aportado por el autor del proyecto, ver `01_comprension_del_negocio.ipynb`): la intención es usar esta variable para construir una clasificación de **"café de especialidad"**, siguiendo el estándar de la industria (SCA) de `Total.Cup.Points >= 80`. Esa construcción de variable objetivo **todavía no existe en el código** y debe implementarse en la fase 03, junto con la decisión sobre el registro con puntaje 0.

## 2.4 TODO — pendiente en esta fase

El notebook original no incluye ninguna visualización. Para cumplir con lo exigido por el profesor (distribuciones, outliers, correlaciones) y con la rúbrica de "Aplicación de CRISP-DM", **falta agregar aquí**:

- [ ] Distribución de `Total.Cup.Points` (y de los subpuntajes: `Aroma`, `Flavor`, `Aftertaste`, `Acidity`, `Body`, `Balance`, etc.) — ej. histograma o boxplot, guardado en `reports/figures/`.
- [ ] Análisis del registro con `Total.Cup.Points == 0` (¿es uno o varios? ¿qué tienen en común?).
- [ ] Matriz de correlación entre los subpuntajes y `Total.Cup.Points`.
- [ ] Distribución de `Country.of.Origin` u otras variables categóricas relevantes para el objetivo de negocio (asesorar al sector cafetero colombiano).
- [ ] Confirmar y documentar la decisión final sobre el desbalance arabica/robusta y sobre el umbral de "café de especialidad".

Las funciones de graficado reutilizables deben implementarse en `src/visualization/plots.py` (hoy vacío a propósito) y llamarse desde aquí, guardando cada figura en `reports/figures/` vía `config.yaml`.